src/java/org/apache/cassandra/db/virtual/PendingHintsTable.java

```diff
@@ -0,0 +1,115 @@
+    private static final String DC = "dc";
+    private static final String STATUS = "status";
+    private static final String FILES = "files";
+    private static final String NEWEST = "newest";
+    private static final String OLDEST = "oldest";
+
+    PendingHintsTable(String keyspace)
+    {
+        super(TableMetadata.builder(keyspace, "pending_hints")
+                           .comment("Pending hints that this node has for other nodes")
+                           .kind(TableMetadata.Kind.VIRTUAL)
+                           .partitioner(new LocalPartitioner(UUIDType.instance))
+                           .addPartitionKeyColumn(HOST_ID, UUIDType.instance)
+                           .addRegularColumn(ADDRESS, InetAddressType.instance)
+                           .addRegularColumn(PORT, Int32Type.instance)
+                           .addRegularColumn(RACK, UTF8Type.instance)
+                           .addRegularColumn(DC, UTF8Type.instance)
+                           .addRegularColumn(STATUS, UTF8Type.instance)
+                           .addRegularColumn(FILES, Int32Type.instance)
+                           .addRegularColumn(NEWEST, TimestampType.instance)
+                           .addRegularColumn(OLDEST, TimestampType.instance)
+                           .build());
+    }
+
+    @Override
+    public DataSet data()
+    {
+        List<PendingHintsInfo> pendingHints = HintsService.instance.getPendingHintsInfo();
+        IEndpointSnitch snitch = DatabaseDescriptor.getEndpointSnitch();
+
+        SimpleDataSet result = new SimpleDataSet(metadata());
+
+        Map<String, String> simpleStates;
+        if (FailureDetector.instance instanceof FailureDetectorMBean)
+            simpleStates = ((FailureDetectorMBean) FailureDetector.instance).getSimpleStatesWithPort();
+        else
+            simpleStates = Collections.emptyMap();
+
+        for (PendingHintsInfo info : pendingHints)
+        {
+            InetAddressAndPort addressAndPort = StorageService.instance.getEndpointForHostId(info.hostId);
+            InetAddress address = null;
+            Integer port = null;
+            String rack = "Unknown";
+            String dc = "Unknown";
+            String status = "Unknown";
+            if (addressAndPort != null)
+            {
+                address = addressAndPort.address;
+                port = addressAndPort.port;
+                rack = snitch.getRack(addressAndPort);
+                dc = snitch.getDatacenter(addressAndPort);
+                status = simpleStates.getOrDefault(addressAndPort.toString(), status);
+            }
+            result.row(info.hostId)
+                  .column(ADDRESS, address)
+                  .column(PORT, port)
+                  .column(RACK, rack)
+                  .column(DC, dc)
+                  .column(STATUS, status)
+                  .column(FILES, info.totalFiles)
+                  .column(NEWEST, new Date(info.newestTimestamp))
```

Comment:
I would use parsing from `Instant` here. Date is kind of ... old. At least in snapshots subsystem we are trying to use new Java time api where it is possible.

Reply:
I've used `Date` here because `TimestampSerializer` used for `TimestampType` type expects it:

https://github.com/apache/cassandra/blob/2e2db4dc40c4935305b9a2d5d271580e96dabe42/src/java/org/apache/cassandra/serializers/TimestampSerializer.java#L41

https://github.com/apache/cassandra/blob/2e2db4dc40c4935305b9a2d5d271580e96dabe42/src/java/org/apache/cassandra/serializers/TimestampSerializer.java#L136